lib/services/fanout.go

```diff
@@ -0,0 +1,217 @@
+		w.cancel()
+		return nil, trace.Wrap(err)
+	}
+	f.addWatcher(w)
+	return w, nil
+}
+
+func filterEventSecrets(event Event) Event {
+	r, ok := event.Resource.(ResourceWithSecrets)
+	if !ok {
+		return event
+	}
+	event.Resource = r.WithoutSecrets()
+	return event
+}
+
+// Emit broadcasts events to all matching watchers that have been attached
+// to this fanout instance.
+func (f *Fanout) Emit(events ...Event) {
+	f.Lock()
+	defer f.Unlock()
+	for _, fullEvent := range events {
+		// by default, we operate on a version of the event which
+		// has had secrets filtered out.
+		event := filterEventSecrets(fullEvent)
+		var remove []*fanoutWatcher
+	Inner:
+		for _, entry := range f.watchers[event.Resource.GetKind()] {
+			match, err := entry.kind.Matches(event)
+			if err != nil {
+				entry.watcher.setError(err)
+				remove = append(remove, entry.watcher)
+				continue Inner
+			}
+			if !match {
+				continue Inner
+			}
+			emitEvent := event
+			// if this entry loads secrets, emit the
+			// full unfiltered event.
+			if entry.kind.LoadSecrets {
+				emitEvent = fullEvent
+			}
+			if err := entry.watcher.emit(emitEvent); err != nil {
+				remove = append(remove, entry.watcher)
+				continue Inner
+			}
+		}
+		for _, w := range remove {
+			f.removeWatcher(w)
+			w.cancel()
+		}
+	}
+}
+
+// CloseWatchers closes all attached watchers, effectively
+// resetting the Fanout instance.
+func (f *Fanout) CloseWatchers() {
+	f.Lock()
+	defer f.Unlock()
+	for _, entries := range f.watchers {
+		for _, entry := range entries {
+			entry.watcher.cancel()
+		}
+	}
+	// watcher map was potentially quite large, so
+	// relenguish that memory.
+	f.watchers = make(map[string][]fanoutEntry)
+}
+
+func (f *Fanout) addWatcher(w *fanoutWatcher) {
+	for _, kind := range w.watch.Kinds {
+		entries := f.watchers[kind.Kind]
+		entries = append(entries, fanoutEntry{
+			kind:    kind,
+			watcher: w,
+		})
+		f.watchers[kind.Kind] = entries
+	}
+}
+
+func (f *Fanout) removeWatcher(w *fanoutWatcher) {
+	for _, kind := range w.watch.Kinds {
+		entries := f.watchers[kind.Kind]
+	Inner:
+		for i, entry := range entries {
+			if entry.watcher == w {
+				entries = append(entries[:i], entries[i+1:]...)
+				break Inner
+			}
+		}
+		switch len(entries) {
+		case 0:
+			delete(f.watchers, kind.Kind)
+		default:
+			f.watchers[kind.Kind] = entries
+		}
+	}
+}
+
+func newFanoutWatcher(ctx context.Context, watch Watch) (*fanoutWatcher, error) {
+	if len(watch.Kinds) < 1 {
+		return nil, trace.BadParameter("must specify at least one resource kind to watch")
+	}
+	ctx, cancel := context.WithCancel(ctx)
+	if watch.QueueSize < 1 {
+		watch.QueueSize = defaultQueueSize
+	}
+	return &fanoutWatcher{
+		watch:  watch,
+		eventC: make(chan Event, watch.QueueSize),
+		cancel: cancel,
+		ctx:    ctx,
+	}, nil
+}
+
+type fanoutWatcher struct {
+	emux   sync.Mutex
+	err    error
+	watch  Watch
+	eventC chan Event
+	cancel context.CancelFunc
+	ctx    context.Context
```

Comment:
would be ideal to not have this as a field
but keep as is if it's too much hassle